### DISCUSSION TITLE
About libraries in OpenMage. Direct usage or Fork?

### DISCUSSION THREAD

**Discussion by addison74** [OpenMage/magento-lts#2430] [General]
As you may know we have in plan to migrate all 3rd party libraries to composer dependencies. There is an open PR for a part of them here #2411.
 
Now please read this post created in a module that we use https://github.com/MyIntervals/emogrifier/issues/1164#issue-1315035805

The author then comes and offers explanations, presenting his point of view why he proposed to drop support for PHP 7.3.

We recently took a step by establishing as the minimum PHP requirement in OpenMage version 7.3, even is EOL we find it in operating systems such as Debian 10 and which will be maintained until June 2024.

The discussion I want to propose is related to these libraries that we want Composer to manage. Until now, OpenMage had all the files in its own repository, but with the modifications we want to make we become dependent on the decisions of those who maintain these projects. 

I think we should consider having our own forks of these projects right here in OpenMage, along with magento-lts. In this way, we can avoid being taken by surprise by radical changes where we do not have immediate solutions.

**Comment by Flyingmana** [comment]
I would say, avoid having a fork of them as long as possible.
And most surprises should be avoidable by setting version constraints which dont allow new major versions to be installed automatically.

**Reply by addison74** [reply]
This seems the best option, to prevent the automatic installation of new versions that could create trouble in relation to the minimum requirements in our project.

**Reply by fballiano** [reply]
I'd also try not to fork unless absolutely necessary.
For example, in PR #2411 I've linked pear/net_idna2 in the "dev-master" version, since the last tagged version didn't include some necessary fix but... it's not a god idea to have a "dev-master" version for sure and I think that project is probably abandoned (I wrote to the maintainer's email address asking to tag a new release but didn't get any answer in some weeks and it's not possible to open issues in that github project) so I'd fork it at least to be able to create a tagged release. Where to fork it tho? under the openmage umbrella or in some dev personal space?

**Reply by addison74** [reply]
I think that a fork should belong to the OpenMage organization and not to a user. As long as the organization exists then that fork will exist too. At least for those on which OpenMage depends to run properly. In addition, certain people will get higher level access to update that fork. I don't think we will implement any immediately, but it is good to be alert to the changes of others we depend on.

**Reply by fballiano** [reply]
I agree that it would be the wisest choice. For pear/net_idna2 I think we have to do it right away but only a high level maintainer can do that.

**Reply by addison74** [reply]
@Flyingmana will help us. You took care of this change and those dependencies in detail. If you consider that the pear/net_idna2 fork is needed then it must be done, even more so if it is abandoned. Maybe we should have a rule that when one of the dependencies is abandoned we should have its fork and change the dependency to it.

**Reply by fballiano** [reply]
actually I discovered something just now that makes it not necessary to fork the net_idna2 repo, it's possible to link a specific commit ID in the composer.json, I just did it here: https://github.com/OpenMage/magento-lts/pull/2411/files#diff-d2ab9925cad7eac58e0ff4cc0d251a937ecf49e4b6bf57f8b95aab76648a9d34 so at least in this case I feel comfortable releasing this specific version of that library :-)

**Reply by fballiano** [reply]
auch sorry guys, my local composer was working fine but the one in the workflows tells that linking to a commit ID is a bad practice:
<img width="1241" alt="Schermata 2022-08-15 alle 20 50 30" src="https://user-images.githubusercontent.com/909743/184707312-856eba1d-daf3-4c7e-ade0-e83b4cf0b335.png">

although I don't see why it should it a bad practise

**Reply by sreichel** [reply]
It seems to be a warning, but the error comes from outdated lock-file.

Edit: :(

> --strict: Return a non-zero exit code for warnings as well as errors

**Reply by fballiano** [reply]
yep, it's the strict validation

**Reply by sreichel** [reply]
I'd use "dev-main", instead of forking just to have a tagged release. This PHP/PEAR lib has no php requirement and won't see any BC breaking changes ... i'm pretty sure.
